databuilder/query_model.py

```diff
@@ -326,23 +347,48 @@ class DomainMismatchError(Exception):
 
 
 def validate_node(node):
-    # TODO: Runtime type validation (i.e. only acceptable types are passed in), possibly
-    # using something like pydantic.
-
-    # TODO: Validation of the "inner" types i.e. a series may contain date values and we
-    # want to make it an error if you try to compare these to an int. And there are
-    # several places (e.g. a Filter condition) where we require suppied series to have
-    # boolean type. We'll need the DSL to get these types from the schema and pass them
-    # in somehow.
-
-    # The one exception to the "common domain" rule is the Join operation which takes
-    # frames from two different domains and produces a new domain. (This is no-cover-ed
-    # because we're not using the Join operation yet.)
-    if not isinstance(node, Join):  # pragma: no cover
-        validate_children_have_common_domain(node)
-
-
-def validate_children_have_common_domain(node):
+    # Check that the types supplied match the types specified
+    validate_types(node)
+    # As well as types we need to validate the "domain constraint". Frames and series
+    # which are in one-row-per-patient form can be combined arbitrarily because we can
+    # JOIN using the patient_id and be sure that we're not creating new rows. But for
+    # operations involving many-rows-per-patient inputs we need to ensure that they are
+    # all drawn from the same underlying table. (We call this the "domain" for set
+    # theoretic reasons which the margin of this comment are too small to contain.) Note
+    # that when we add a Join operation that will be the one explicit exception to this
+    # rule.
+    validate_inputs_have_common_domain(node)
+
+
+def validate_types(node):
+    # Within the context of this node, any TypeVars evaluated must take consistent
+    # values so we created a single context object to share between all fields
+    typevar_context = {}
+    for field in dataclasses.fields(node):
```

Comment:
I have actually been thinking of this as a generic dataclass type checker that happens to only care about the types we use in the query model. Obviously all the hard work is with Series objects, but checking that things which are supposed to be strings are actually strings (rather than e.g some contract class) seems like it might still be a useful thing to do, and given that it would be more effort to _not_ check them I think we should.

Reply:
Yup. Do you think it's worth doing a little more work to make this _obviously_ generic and them moving it out of this module? (To reduce the cognitive load -- allow people working on the query model to take it as read that this is a dataclass-focused thing with no query model cleverness in it.)

Reply:
I think it falls in the bucket of "things that require slightly too much thinking to do right now" — though I do agree that the more stuff we can make generic and move out the better.